### PR TITLE
Jackett "Standalone" aka build using .NET

### PR DESCRIPTION
**Pull request by publicarray**
_Motivation:_  https://github.com/SynoCommunity/spksrc/issues/3697#issuecomment-698116461
_Linked issues:_ #3697

CC @m4tt075

**Caveats:**
* Only `x64`, `ARM 8` and `ARM 7` are supported, ~~so I've moved it to a new package: `jackett-dotnet`~~
~~* Currently the .NET framework keeps data in `~/.dotnet`, so I used `docker run --name spksrc --rm -it -v $HOME/git/spksrc:/spksrc:delegated -v :/root/.dotnet synocommunity/spksrc:latest /bin/bash` to run docker~~

Build: https://github.com/publicarray/spksrc/releases/tag/jackett-dotnet-0.16.1583

### Checklist
- [x] Build rule `all-supported` completed successfully
- [x] Package upgrade completed successfully
- [x] New installation of package completed successfully

### CODE REVIEW

**Comment by m4tt075** [comment]
@publicarray Thanks for copying me in. I've stopped using Synology NASs some time ago, and as such, am no longer maintaining this package. As such please feel free to put yourself or just SynoCommunity in as maintainer.

In any case, I remember very vividly that maintaining mono has been more than painful. As such, this PR and your efforts here to circumvent the mono dependency will clearly be beneficial and appreciated by the community!

**Comment by publicarray** [comment]
Oh I am curious, Can I ask what you are using instead for storage?

Thanks for your kind words 🙏

**Comment by m4tt075** [comment]
I've built my own NAS. If you are interested in the background and you understand (some) German, you can find the whole story here: https://www.kodinerds.net/index.php/Thread/65922-Neues-NAS-Server/?pageNo=1.

**Comment by publicarray** [comment]
@m4tt075 Thank you! That is very interesting. (Ich kann Deutsch sprechen 😄)

**Review by ymartin59 — CHANGES_REQUESTED** [review]
I recommend to replace original "jackett" (with Mono runtime) with new version compiled by dotNetCore and to confirm upgrade process allows user to keep application configuration and data.

**Comment by publicarray** [comment]
@ymartin59 Done, I replaced the original jackett. Note that some older 32-bit architectures are not supported. I've also created generic makefiles for .net

**Comment by ymartin59** [comment]
@publicarray So it means package is no longer available for "evansport" whereas this architecture is still supported with DSM 7? It is quite annoying. What prevents i686 compilation to work with dotnet-core?

**Comment by hgy59** [comment]
> 
> 
> @publicarray So it means package is no longer available for "evansport" whereas this architecture is still supported with DSM 7? It is quite annoying. What prevents i686 compilation to work with dotnet-core?

There is no support for 32 bit linux with .NET (formerly named dotnet core). 32 bit is supported only on ARM (linux arm32) and windows (x86).
see [current versions](https://dotnet.microsoft.com/download/dotnet/5.0)

File: native/dotnet-sdk-5.0/Makefile [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+INSTALL_TARGET = dotnet-install
+
+DOTNET_PACKAGES=$(DISTRIB_DIR)/nuget/packages
+
+include ../../mk/spksrc.native-cc.mk
+
+# Supported architectures:
+# https://github.com/dotnet/core/blob/master/release-notes/3.1/3.1-supported-os.md#linux
+## UNSUPPORTED_ARCHS = $(PPC_ARCHES) $(ARM5_ARCHES) $(x86_ARCHES)
+
```
Comment (ymartin59) [inline native/dotnet-sdk-5.0/Makefile:25]: Please discard these lines. Not relevant for "native" tool.

File: native/dotnet-sdk-5.0/Makefile [hunk 2]
```diff
@@ -0,0 +1,28 @@
+PKG_NAME = dotnet-sdk-5.0
+PKG_VERS = 5.0.100
+PKG_EXT = tar.gz
+PKG_DIST_SITE = https://dotnetcli.azureedge.net/dotnet/Sdk/${PKG_VERS}
+PKG_DIST_NAME = dotnet-sdk-$(PKG_VERS)-linux-x64.$(PKG_EXT)
+
+# https://docs.microsoft.com/en-us/dotnet/core/install/dependencies?pivots=os-linux&tabs=netcore31
+DEPENDS =
+
```
Comment (ymartin59) [inline native/dotnet-sdk-5.0/Makefile:9]: Please discard these lines. Not relevant for "native" tool.

File: native/dotnet-sdk-3.1/Makefile [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+# https://github.com/dotnet/core/blob/master/release-notes/3.1/3.1.9/3.1.9.md
+PKG_VERS = 3.1.404
+PKG_EXT = tar.gz
+# https://download.visualstudio.microsoft.com/download/pr/ec187f12-929e-4aa7-8abc-2f52e147af1d/56b0dbb5da1c191bff2c271fcd6e6394/dotnet-sdk-3.1.404-linux-x64.tar.gz
+PKG_DIST_SITE = https://dotnetcli.azureedge.net/dotnet/Sdk/${PKG_VERS}
+PKG_DIST_NAME = dotnet-sdk-$(PKG_VERS)-linux-x64.$(PKG_EXT)
+
+# https://docs.microsoft.com/en-us/dotnet/core/install/dependencies?pivots=os-linux&tabs=netcore31
+DEPENDS =
+
```
Comment (ymartin59) [inline native/dotnet-sdk-3.1/Makefile:12]: Please discard these lines. Not relevant for "native" tool.

File: native/dotnet-sdk-3.1/Makefile [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+INSTALL_TARGET = dotnet-install
+
+DOTNET_PACKAGES=$(DISTRIB_DIR)/nuget/packages
+
+include ../../mk/spksrc.native-cc.mk
+
+# Supported architectures:
+# https://github.com/dotnet/core/blob/master/release-notes/3.1/3.1-supported-os.md#linux
+## UNSUPPORTED_ARCHS = $(PPC_ARCHES) $(ARM5_ARCHES) $(x86_ARCHES)
+
```
Comment (ymartin59) [inline native/dotnet-sdk-3.1/Makefile:28]: Please discard these lines. Not relevant for "native" tool.

**Review by ymartin59 — CHANGES_REQUESTED** [review]
Looks good. But I would request effort to keep Jackett for archs not supported by dotnetcore, thanks to a conditional build with Mono. At least for archs still under support with DSM 7...

File: spk/jackett/Makefile [hunk 5]
```diff
@@ -28,14 +26,9 @@ SERVICE_PORT_TITLE = Jackett(HTTP)
 # Admin link
 ADMIN_PORT = $(SERVICE_PORT)
 
-POST_STRIP_TARGET = jackett_extra_install
-
-REQUIRED_DSM = 5.0
-# no PPC support for mono, see cross/mono/Makefile for details
-UNSUPPORTED_ARCHS = $(PPC_ARCHS)
+UNSUPPORTED_ARCHS = $(PPC_ARCHS) $(ARMv5_ARCHS) $(i686_ARCHS) $(ARMv7L_ARCHS)
```
[the remark this answers is not in the document]
Comment (publicarray) [reply spk/jackett/Makefile]: I gave it a go, let me know what you think.

File: spk/jackett/Makefile [hunk 6]
```diff
@@ -1,19 +1,17 @@
...
 SPK_ICON = src/jackett.png
 DSM_UI_DIR = app
 
-DEPENDS = cross/busybox cross/curl cross/$(SPK_NAME)
-# depend on mono 5.4 for BTLS HttpWebClient support
-SPK_DEPENDS = "mono>=5.4"
+DEPENDS = cross/$(SPK_NAME)
 
-MAINTAINER = m4tt075
+MAINTAINER = SynoCommunity
```
[the remark this answers is not in the document]
Comment (publicarray) [reply spk/jackett/Makefile:9]: I rather not show up with my username in software that is most likely used for piracy.

File: spk/jackett/Makefile [hunk 7]
```diff
@@ -1,19 +1,17 @@
...
 DESCRIPTION = "Jackett works as a proxy server. It translates queries from apps into tracker-site-specific http queries, parses the html response, then sends results back to the requesting software. Jackett is a single repository of maintained indexer scraping and translation logic - removing the burden from other apps."
 RELOAD_UI = yes
 DISPLAY_NAME = Jackett
 STARTABLE = yes
-CHANGELOG = "Add workaround for mono bug with armv5/88f6281"
+CHANGELOG = "Compile using .NET Core"
```
Comment (ymartin59) [inline spk/jackett/Makefile:14]: I would manage ChangeLog conditionally according to target arch... either mono or dotnetcore

File: spk/jackett/Makefile [hunk 8]
```diff
@@ -28,14 +26,24 @@ SERVICE_PORT_TITLE = Jackett(HTTP)
...
+SERVICE_SETUP = src/service-setup-mono.sh
+endif
+ifeq ($(findstring $(ARCH),$(i686_ARCHS)),$(ARCH))
+DEPENDS = cross/$(SPK_NAME)-mono
+SERVICE_SETUP = src/service-setup-mono.sh
+endif
+ifeq ($(findstring $(ARCH),$(ARMv7L_ARCHS)),$(ARCH))
+DEPENDS = cross/$(SPK_NAME)-mono
+SERVICE_SETUP = src/service-setup-mono.sh
+endif
```
Comment (ymartin59) [inline spk/jackett/Makefile:44]: Same statements. Please use a single "ifeq" condition with all of them, and include ChangeLog here

**Review by ymartin59 — CHANGES_REQUESTED** [review]
Thanks for this effort. Looks good.
Does it really make sense to include dotnet-sdk-3.1 here? You may keep files aside for you/future applications, but I guess there is little chance an application remains on old sdk version?
Sorry but I still have some minor improvements. Then I will test this "branching" package as I have evansport.

**Comment by publicarray** [comment]
jellyfin is still on 3.1 (at least until 10.7 is released). But generally agree. 
3.1 is almost supported for another 2 years (2022-12-03). So I would keep it just in case for now.

**Review by ymartin59 — APPROVED** [review]

**Comment by Wuszek** [comment]
After upgrading on DS218+, couldn't start package. Had to manually downgrade to last working version.

**Comment by publicarray** [comment]
Thanks @Wuszek. Sigh. Why is it never simple... Are there any logs?

**Comment by Wuszek** [comment]
Where do I look @publicarray ? Now I have installed Jackett from the previous release (and this mono version installed and started fine), but maybe logs are still somewhere over there

**Comment by publicarray** [comment]
Thanks @Wuszek, Sorry do you have a log file at `cat /var/packages/jackett/target/var/jackett.log` ?

**Comment by Wuszek** [comment]
@publicarray ok, so I just upgraded Jackett in Package Center and I get "Couldnt start up package" notification. This is jackett.log:

Sun Feb 21 13:21:56 CET 2021
Starting jackett command env HOME=/volume1/@appstore/jackett/var PATH=/volume1/@appstore/jackett/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/syno/sbin:/usr/syno/bin:/usr/local/sbin:/usr/local/bin LD_LIBRARY_PATH=/volume1/@appstore/jackett/lib /volume1/@appstore/jackett/share/jackett --PIDFile /volume1/@appstore/jackett/var/jackett.pid 
/volume1/@appstore/jackett/share/jackett: /lib64/libstdc++.so.6: no version information available (required by /volume1/@appstore/jackett/share/jackett)
/volume1/@appstore/jackett/share/jackett: /lib64/libstdc++.so.6: no version information available (required by /volume1/@appstore/jackett/share/jackett)
/volume1/@appstore/jackett/share/jackett: /lib64/libstdc++.so.6: no version information available (required by /volume1/@appstore/jackett/share/jackett)
/volume1/@appstore/jackett/share/jackett: /lib64/libstdc++.so.6: no version information available (required by /volume1/@appstore/jackett/share/jackett)
/volume1/@appstore/jackett/share/jackett: /lib64/libstdc++.so.6: no version information available (required by /volume1/@appstore/jackett/share/jackett)
/volume1/@appstore/jackett/share/jackett: /lib64/libstdc++.so.6: no version information available (required by /volume1/@appstore/jackett/share/jackett)
/volume1/@appstore/jackett/share/jackett: relocation error: /volume1/@appstore/jackett/share/jackett: symbol _ZNSt25__codecvt_utf8_utf16_baseIDsED2Ev, version GLIBCXX_3.4.21 not defined in file libstdc++.so.6 with link time reference

**Comment by publicarray** [comment]
Thanks that explains it clearly. The glibc is too old. The x64 I'm testing on has a slightly newer version.

**Comment by Wuszek** [comment]
Is there a command to check which version I am using? I tried
`strings /usr/lib/libstdc++.so.6 | grep LIBCXX`

but there is no strings command

**Comment by Wuszek** [comment]
@publicarray ping me pls when there would be fix for this, Im willing to test this

**Comment by publicarray** [comment]
Thanks @Wuszek what is the output of `ls -l /usr/lib/libstdc++.so.6` ?

```
ls -l /lib64/libc.so.6
lrwxrwxrwx 1 root root 20 Dec 14 17:36 /lib64/libc.so.6 -> libc-2.20-2014.11
ls -l /lib64/libstdc++.so.6
lrwxrwxrwx 1 root root 19 Dec 14 17:36 /usr/lib/libstdc++.so.6 -> libstdc++.so.6.0.20
```

As a workaround you can copy the lib from a linux box:

```
$ scp /lib64/libstdc++.so.6 user@synology:/tmp/libstdc++.so.6
$ ssh user@synology
# sudo mv /tmp/libstdc++.so.6 /var/packages/jackett/target/lib/libstdc++.so.6
```

I've deactivated the .NET versions of the package for now

**Comment by Wuszek** [comment]
@publicarray output:

```
root@SynologyNAS_WK:~# ls -l /usr/lib/libstdc++.so.6
lrwxrwxrwx 1 root root 19 May 26  2020 /usr/lib/libstdc++.so.6 -> libstdc++.so.6.0.20
```

**Comment by ta264** [comment]
@Wuszek We're also very keen to finalize the switch away from mono for Radarr... Would you be up for joining the radarr discord and trying out a few ideas? https://radarr.video/discord

**Pull request merged**